### PR TITLE
analyzer/runtime: Index (un)delegate events, txs

### PR DESCRIPTION
**Pull request by mitjat**
Adds indexing of txs that call the `consensus.Delegate` and `consensus.Undelegate` methods, and of `consensus_accounts` SDK events of type `Delegate`, `UndelegateStart`, `UndelegateDone`.

TODO: Testing. For starters, I'll run this on the last few weeks of data on testnet, which is where these txs and events first appeared.

### CODE REVIEW

File: analyzer/runtime/visitors.go
Comment (mitjat) [inline analyzer/runtime/visitors.go:1]: The changes in this file are purely boilerplate.

File: analyzer/runtime/extract.go
Comment (mitjat) [inline analyzer/runtime/extract.go:385]: This is analogous to (and copy pasted from) our existing handling of `consensusaccounts.Withdraw` calls

File: analyzer/runtime/extract.go
Comment (mitjat) [inline analyzer/runtime/extract.go:398]: Can anybody confirm this?

Also, if true, should we store the number of shares in the `amount` variable even though the semantics are different? Both `to` and `amount` are loosely defined and introduced by the indexer; they're a kind of best-effort intentionally-over-generalized fields to help navigate the txs quickly almost regardless of their call method.

The `amount` variable contents end up in the `amount` column in our DB, and eventually in the `amount` field of each tx in the API, but the API docs describe the above looseness pretty well:
> A reasonable "amount" associated with this transaction, if applicable. The meaning varies based on the transaction method.

Still, might be too much of a stretch? And potentially particularly confusing because shares and token amounts are so closely related?

File: analyzer/runtime/extract.go
Comment (mitjat) [inline analyzer/runtime/extract.go:395]: While the `to` address according to tx-internal semantics is indeed the tx sender (see NOTE comment above), it might make more sense to store `body.From` in our `to` field? The `to` field is internally defined by the indexer (though exposed through the API), with [varying meanings](https://github.com/oasisprotocol/nexus/blob/2ea104969cc4242ca896a7d50bd313024c849432/api/spec/v1.yaml#L2368-L2375) but generally being "the address that is most relevant to this tx, other than the sender".

File: analyzer/runtime/extract.go
Comment (mitjat) [inline analyzer/runtime/extract.go:584]: Unrelated simplification

File: analyzer/runtime/extract.go
Comment (mitjat) [inline analyzer/runtime/extract.go:643]: I did not follow this pattern with (un)delegate events, and I'm tempted to undo it for these "old" events to, i.e. I'm tempted to always extract `.To` as a related address. There's not really a downside to overindexing, other than bloating our DB a bit, as far as I know. And it might come handy to be able to look up `consensusaccounts` events related to a certain consensus account.

File: analyzer/runtime/extract.go
Comment (mitjat) [inline analyzer/runtime/extract.go:379]: Can someone confirm this please? `to` is typically a validator account, right? And if one delegated to a non-validator account, nothing bad would happen, it just wouldn't result in any rewards over time.

Comment (kostko) [reply analyzer/runtime/extract.go:379]: Correct for both.

File: analyzer/runtime/extract.go
Comment (kostko) [reply analyzer/runtime/extract.go:398]: Do not conflate shares and amount as this can lead to incorrect values and user confusion.

You can only know the final amount once the undelegation completes as nodes may be slashed while debonding. However you can compute the amount at any point in time by inspecting the relevant share pool (see the CLI).

File: analyzer/runtime/extract.go
Comment (pro-wh) [reply analyzer/runtime/extract.go:385]: :+1:

File: analyzer/runtime/extract.go
Comment (pro-wh) [reply analyzer/runtime/extract.go:643]: ya ok

File: analyzer/runtime/extract.go
Comment (pro-wh) [reply analyzer/runtime/extract.go:395]: :skull: is this what we do for consensus undelegate?

**Review by pro-wh — COMMENTED** [review]
lg but let's approve when we're sure there will be no regrets about what "to" in an undelegate means

File: analyzer/runtime/extract.go
Comment (mitjat) [reply analyzer/runtime/extract.go:395]: By "consensus undelegate", you mean the `ReclaimEscrow` method from [here](https://github.com/oasisprotocol/oasis-core/blob/6dcc5f29ff908f8c6c5bd3481f20969db009c3be/go/staking/api/api.go#L77-L81), right?
When [inserting consensus txs into the DB](https://github.com/oasisprotocol/nexus/blob/2ea104969cc4242ca896a7d50bd313024c849432/analyzer/consensus/consensus.go#L332-L345), we make no attempt to construct a reasonably generalizable `to` field. So I guess there's no prior art :/

I'm actually inclined to set `to = body.From`. Because otherwise there's no added value in setting the `to` (we're storing `body.To` in the tx sender field already), and so it will be hard to surface undelegations relating to a specific validator.

Edit: Doing so after a slack discussion.

File: analyzer/runtime/extract.go
Comment (mitjat) [reply analyzer/runtime/extract.go]: Thanks. I kept the shares in the `amount` field, as was coded initially; I agree we better be careful here.
I updated the comment with your explanation about slashing.

File: analyzer/runtime/extract.go
Comment (mitjat) [reply analyzer/runtime/extract.go]: After more discussion on Slack, we can't think of reasons why this would be problematic. So I'm adding `.To` (for `Withdraw` event) and `.From` (for `Deposit`) to the list of event-related addrs.

**Comment by mitjat** [comment]
After discussions, I slightly tweaked what addresses we index. Reviewers, please take another look. Thanks!

File: analyzer/runtime/extract.go
Comment (ptrus) [reply analyzer/runtime/extract.go:376]: I tnink no need, since the appropriate `account` burn/mint events will be emitted by the SDK on the delegation and once the funds are reclaimed. And those are already handled in https://github.com/oasisprotocol/nexus/blob/fb5d2fb1ed3836f0bf15d2edd076fd8b255ba9e6/analyzer/runtime/accounts.go#L18-L31

File: analyzer/runtime/extract.go
Comment (ptrus) [reply analyzer/runtime/extract.go:398]: I think that no need, since SDK will also emit a `Mint` event on the undelegation: https://github.com/oasisprotocol/oasis-sdk/blob/fac9e70f9444e0f20e62ca56cc597abfd3f54e38/runtime-sdk/src/modules/consensus_accounts/mod.rs#L788

File: analyzer/runtime/extract.go
Comment (ptrus) [inline analyzer/runtime/extract.go:376]: This should probably check if the event contains an error https://github.com/oasisprotocol/oasis-sdk/blob/fac9e70f9444e0f20e62ca56cc597abfd3f54e38/client-sdk/go/modules/consensusaccounts/types.go#L201


Actually all `consensus_account` events can contain a potential consensus error in which indicates a failed event, and should probably be skipped here?

See: https://github.com/oasisprotocol/oasis-sdk/blob/fac9e70f9444e0f20e62ca56cc597abfd3f54e38/client-sdk/go/modules/consensusaccounts/types.go#L233-L239

and: https://github.com/oasisprotocol/oasis-sdk/blob/fac9e70f9444e0f20e62ca56cc597abfd3f54e38/runtime-sdk/src/modules/consensus_accounts/mod.rs#L609-L625

for examples.

Comment (mitjat) [reply analyzer/runtime/extract.go:374]: Correct. (With a small caveat -- there's an additional Transfer involved.) I was also unsure of this when I started this PR, and it took me some spelunking to find the explanation before I was comfortable writing the code. Good question, Andy, and worth spelling out the answer for future maintainers. See the new "LESSON" comment in the code for an explanation of events.

Comment (mitjat) [reply analyzer/runtime/extract.go:374]: Hmm ... it does feel weird to basically ignore errors, but I'm not sure the indexer needs to do anything special about `consensusaccounts` events with a non-nil `Error`:
- There is no dead reckoning to adjust (or suppress); that all happens via other, co-emitted events
- We [store](https://github.com/oasisprotocol/nexus/blob/f9aae40e235f3dca08bb2845f8fb124603b8f310/analyzer/runtime/extract.go#L397-L397) (and then expose via API) the entire body of the event as JSON; [example](https://index-staging.oasislabs.com/v1/emerald/events?type=consensus_accounts.deposit). If the `Error` field were set, we'd include it there.

Maybe I'm forgetting something?

Comment (ptrus) [reply analyzer/runtime/extract.go:374]: You're right

**Review by ptrus — APPROVED** [review]

File: analyzer/runtime/extract.go
Comment (Andrew7234) [reply analyzer/runtime/extract.go:374]: thanks for adding the comnent! v helpful

**Review by Andrew7234 — APPROVED** [review]

**Pull request merged**
